src/config.rs

```diff
@@ -14,8 +23,141 @@ impl Config {
     }
 }
 
-#[derive(Debug, Deserialize)]
+#[derive(Clone, Debug)]
+pub struct LocalizedConfig {
+    pub config: Config,
+    /// The directory containing `./.cargo/config.toml`
+    pub workspace: PathBuf,
+}
+
+impl Deref for LocalizedConfig {
+    type Target = Config;
+
+    fn deref(&self) -> &Self::Target {
+        &self.config
+    }
+}
+
+impl LocalizedConfig {
+    pub fn new(workspace: PathBuf) -> Result<Self, Error> {
+        Ok(Self {
+            config: Config::parse_from_toml(&workspace.join(".cargo/config.toml"))?,
+            workspace,
+        })
+    }
+
+    /// Search for `.cargo/config.toml` in any parent of the workspace root path.
+    /// Returns the directory which contains this path, not the path to the config file.
+    fn find_cargo_config_parent(workspace: impl AsRef<Path>) -> Result<Option<PathBuf>, Error> {
+        let workspace = dunce::canonicalize(workspace)?;
+        Ok(workspace
+            .ancestors()
+            .find(|dir| dir.join(".cargo/config.toml").is_file())
+            .map(|p| p.to_path_buf()))
+    }
+
+    /// Search for and open `.cargo/config.toml` in any parent of the workspace root path.
+    pub fn find_cargo_config_for_workspace(
+        workspace: impl AsRef<Path>,
+    ) -> Result<Option<Self>, Error> {
+        let config = Self::find_cargo_config_parent(workspace)?;
+        config.map(LocalizedConfig::new).transpose()
+    }
+
+    pub fn resolve_env(&self, key_name: &str) -> io::Result<Option<Cow<'_, str>>> {
+        self.config
+            .env
+            .get(key_name)
+            .map(|v| v.resolve_value(key_name, &self.workspace))
+            .transpose()
+    }
```

Comment:
@dvc94ch @msiglreith Noticed that this only falls back to `std::env::var` if the variable is set in `[env]` but `force=false`. I think we should either:

- Also fall back if the variable is not configured in `[env]` (ie. `.get()` here returns `None`);
- Never fall back, but return an `Option<(Cow..., bool)>` where the `bool` represents the `force` flag. Then the caller can decide on the proper `std::env::var` semantics. Thoughts?

Reply:
I documented this odd behaviour for now, but I don't think we should roll with it :)

Reply:
Fixed this with the help of some test-driven-development :grimacing: